### ISSUE TITLE
Can you explain more about no_aug_epochs args ?

### ISSUE THREAD

**Issue by RTae** [Megvii-BaseDetection/YOLOX#555]
I read your paper and got the no_aug_epochs is use to close the mosaic augment. When I look to the code, I got this part for close mosaic and add l1 loss to training. 
```
def before_epoch(self):
    logger.info("---> start train epoch{}".format(self.epoch + 1))

    if self.epoch + 1 == self.max_epoch - self.exp.no_aug_epochs or self.no_aug:
        logger.info("--->No mosaic aug now!")
        self.train_loader.close_mosaic()
        logger.info("--->Add additional L1 loss now!")
        if self.is_distributed:
            self.model.module.head.use_l1 = True
        else:
            self.model.head.use_l1 = True
        self.exp.eval_interval = 1
        if not self.no_aug:
            self.save_ckpt(ckpt_name="last_mosaic_epoch")
```
I have two questions.
First, Why do you not use mosaic for entries training?  
Second, in the paper, it's said you use **BCE Loss** and **IoU Loss**. Why do you add **l1 loss** to training after close mosaic?

**Comment by Joker316701882** [comment]
@RTae 
1. Mosaic will introduce tons of inaccurate bounding boxes. You can refer to a sample image here:
https://github.com/ultralytics/yolov5/issues/2151
So close the Mosaic will avoid those inaccurate bounding boxes. We have trained with mosaic for 285 epochs, we can still enjoy the gain of Mosaic.

2. Same reason as mentioned above. Before closing the aug, there are many inaccurate bounding boxes, so using too strong a regression loss will hurt the performance. So we add l1 loss only for the last 15 epochs. But this will only improve the mAP by 0.1~0.2%. Also, you can replace IoU Loss with GIoU Loss after closing mosaic, which will probably give you 0.1% more gain. But overall, these are very trival. So we didn't mention them in our tech report.

**Comment by RTae** [comment]
@Joker316701882 Thank you, your answer help me a lot to understand yolox. Last, I have some favor to ask you. Do you have any source about anchor-free detector?

**Comment by glenn-jocher** [comment]
@Joker316701882 I'm experimenting with closing mosaic on YOLOv5 like you have with YOLOX, and was wondering if you tried other values in addition to 15, i.e. 30 epochs or 10 epochs?

I was thinking the EMA is in the middle here confusing the close-mosaic effect since changes to the base model will take many iterations to begin to propagate into the EMA, but I don't see any action to take there.

**Comment by Joker316701882** [comment]
@glenn-jocher Hi. In our experiments, we can observe an immediate improvement on mAP after closing mosaic_aug. It usually achieves the best result between the 5th and 10th epoch after that. We set the default no_aug_epochs to 15 to ensure sufficient iterations on some small customed datasets.

On COCO, it performs similarly when no_aug_epochs is set to 10 and 15. But 5 is a bit worse than 10 and 15 epochs.

**Comment by glenn-jocher** [comment]
@Joker316701882 ok super interesting, thanks! I just started two trainings closing mosaic at 15 and 30, I'll report back my results here.

**Comment by glenn-jocher** [comment]
@Joker316701882 ok my two runs are done, I see the same results as you, immediate improvement for about 5-10 epochs and then a drop.

`--close-mosaic 15` peaked 8 epochs after closing, -`-close-mosaic 30` peaked 10 epochs after closing. This was at --batch-size 128 on COCO for 300 epochs. Both showed about +0.3 mAP bump :)

**Comment by jtorhoff** [comment]
In one of my experiments on a custom dataset I've seen an immediate improvement of mAP after one epoch of no mosaic + L1 then a slow steady decline of both mAP and recall after every following epoch. It's possible that the dataset just needs more "regular" epochs, though, I'm not sure. 

@Joker316701882 Have you encountered this behavior in one of your experiments? Would you consider this "normal" for some datasets or do you think there is some issue? 

Just wondering about you intuition, I would try more experiments myself with e.g. more epochs but it takes a lot of time with limited hardware, so wanted to ask you first.

Also wonder if you tried averaging the weights of the last epochs, i.e. Polyak Averaging or Stochastic Weight Averaging in your experiments.
